gapic-generator-java/src/main/java/com/google/api/generator/gapic/protoparser/Parser.java

```diff
@@ -1124,4 +1158,27 @@ static String parseNestedProtoTypeName(String fullyQualifiedName) {
             .collect(Collectors.toList());
     return String.join(".", nestedTypeComponents);
   }
+
+  /**
+   * Converts a serviceYaml file to a map of methods and autopopulated fields. Note: this does NOT
+   * currently support wildcards in MethodSettings.selectors.
+   */
+  @VisibleForTesting
+  static Map<String, List<String>> parseAutoPopulatedMethodsAndFields(
+      Optional<com.google.api.Service> serviceYamlProtoOpt) {
+    if (!hasMethodSettings(serviceYamlProtoOpt)) {
+      return ImmutableMap.<String, List<String>>builder().build();
+    }
+    return serviceYamlProtoOpt.get().getPublishing().getMethodSettingsList().stream()
+        .collect(
+            Collectors.toMap(
+                MethodSettings::getSelector, MethodSettings::getAutoPopulatedFieldsList));
+  }
+
+  @VisibleForTesting
+  static boolean hasMethodSettings(Optional<com.google.api.Service> serviceYamlProtoOpt) {
+    return serviceYamlProtoOpt.isPresent()
+        && serviceYamlProtoOpt.get().hasPublishing()
+        && !serviceYamlProtoOpt.get().getPublishing().getMethodSettingsList().isEmpty();
```

Comment:
Could `getMethodSettingsList()` return null? If it is, I think we can change this to a null check. We don't have to check if it's empty because `Collectors.toMap()` should map it to an empty map if `getMethodSettingsList()` is empty.

Reply:
I think it will never be null because the default is `java.util.Collections.emptyList()`.